### ISSUE TITLE
mxfp4 MoE decode GEMM: two regressions on gfx942 (3.7.1 trackers flag, and 3.8.0)

### ISSUE THREAD

**Issue by afriedri** [ROCm/triton#970]
# mxfp4 MoE decode GEMM: two regressions on gfx942 (3.7.1 trackers flag, and 3.8.0)

## Summary

The mxfp4 MoE GEMM at **decode shapes** (M=1) has regressed twice on MI300X, from two
independent causes. Against a triton 3.6.0 baseline of ~82 us for the same kernel:

| triton | trackers flag | us/call | vs 3.6 |
|---|---|---|---|
| 3.6.0 | not in tree | 82.28 | — |
| 3.7.1 | OFF | 80.40 | +2% |
| 3.7.1 | **ON** (upstream default) | **121.23** | **-47%** |
| 3.8.0 @ `669b31ac` (tip) | **ON** (upstream default) | **134.55** | **-63%** |

**Regression 1 — 3.7.1 + `amdgpu-use-amdgpu-trackers`.**
[#961 / #11285](https://github.com/ROCm/triton/pull/961) enables the flag unconditionally for
gfx942/gfx950. Toggling it alone accounts for the whole 3.7.1 gap (121 -> 80 us). Worth ~20%
end-to-end tokens/s on gpt-oss-20b.

**Regression 2 — 3.8.0, independent of the flag.**
At tip of the 3.8 branch (`669b31acc1dd1b3fd93286afd5db67f65d9f7557`, 2026-09-09) the same
kernel is ~63% slower than 3.6 — worse than *either* 3.7.1 arm, including 3.7.1 with the
trackers flag on. This is a second, unrelated regression: fixing the flag does not recover
it, and upgrading to 3.8 does not recover regression 1.

Both are decode-shape-specific: the prefill-shape sibling `64x256x128x1` is unaffected
(577 vs 581 us). In end-to-end traces the decode MoE GEMMs are 98.7% of the total loss
(25.25 ms of 25.59 ms over 31 iterations); every other kernel is flat, TTFT is unchanged,
and the entire loss lands in TPOT.

## Affected kernel

`_matmul_ogs_NNT_bf16xbf16xmxfp4_32x256x128x1` and its `_swiglu` variant, from
`python/triton_kernels`. Config: 32x256x128 tile, `num_warps=2`, `num_stages=2`,
`shared=25600`.

## Reproducer

Script covers both regressions — it just reports the kernel's own GPU time, so run it
under whichever triton you want to compare. ~30 s per arm.

### Regression 1 (3.7.1 trackers) — runs on the stock published image, no build

```
docker run --rm -it --device=/dev/kfd --device=/dev/dri --network=host \
  --group-add video --ipc=host --shm-size 16G \
  vllm/vllm-openai-rocm:v0.28.0 bash
```

That image ships triton `3.7.1+gitf0b55c07` — the commit that turned the flag on — and
bundles `triton_kernels`. Inside the container:

```bash
export PYTHONPATH=/usr/local/lib/python3.12/dist-packages/vllm/third_party

rm -rf ~/.triton/cache && python repro_min.py                 # flag ON  (upstream default)
rm -rf ~/.triton/cache && NO_TRACKERS=1 python repro_min.py   # flag OFF
```

```
triton 3.7.1  gfx942:sramecc+:xnack-  NO_TRACKERS=0
  _matmul_ogs_NNT_bf16xbf16xmxfp4_32x256x128x1.kd    121.23 us x100

triton 3.7.1  gfx942:sramecc+:xnack-  NO_TRACKERS=1
  _matmul_ogs_NNT_bf16xbf16xmxfp4_32x256x128x1.kd     80.40 us x100
```

`NO_TRACKERS` is **not** an upstream knob — it's a one-line gate so the two arms can be
compared. In `third_party/amd/backend/compiler.py`, `make_amdgcn()` (~L471 on f0b55c07):

```python
 if options.arch in ["gfx942", "gfx950"]:
     flags.append("amdgpu-use-amdgpu-trackers")
```
becomes
```python
 import os
 if options.arch in ["gfx942", "gfx950"] and os.environ.get("NO_TRACKERS", "0") != "1":
     flags.append("amdgpu-use-amdgpu-trackers")
```

In the stock image that file is at
`/usr/local/lib/python3.12/dist-packages/triton/backends/amd/compiler.py`.

### Regression 2 (3.8.0) — same script, 3.8 installed over the same image

We are not aware of a published image that ships 3.8, so this arm needs triton built from
source into the `v0.28.0` container above. Measured at tip of the 3.8 branch,
`669b31acc1dd1b3fd93286afd5db67f65d9f7557` (2026-09-09). **No patch needed** — this is
stock upstream. Same script, same shape, 3 reps:

```
triton 3.8.0  gfx942:sramecc+:xnack-
  _matmul_ogs_NNT_bf16xbf16xmxfp4_32x256x128x1.kd    134.85 / 134.17 / 134.64 us
```

vs ~82 us on 3.6.0 for the same kernel. The trackers flag is not the bottleneck here: this
commit does pass it on gfx942, but gating it off only recovers ~2% of the 63% gap.

For a 3.6 baseline (82.28 us), the same script under triton 3.6.0, which does not have the
flag in-tree.

<details>
<summary><b>repro_min.py</b> (63 lines)</summary>

```python
"""
Minimal repro: `amdgpu-use-amdgpu-trackers` slows the mxfp4 MoE GEMM ~1.5x at
decode shapes (M=1) on gfx942 / MI300X.

Single matmul call. Only triton + triton_kernels (both ship in the Triton repo).

Run both arms, clearing the cache between them:
    rm -rf ~/.triton/cache && python repro_min.py                # flag ON  (default)
    rm -rf ~/.triton/cache && NO_TRACKERS=1 python repro_min.py  # flag OFF (needs patch)

Expected on MI300X: ~119 us (ON) vs ~79 us (OFF).
"""
import os, torch, triton
from torch.profiler import profile, ProfilerActivity
from triton_kernels.matmul_ogs import (
    matmul_ogs, PrecisionConfig, FlexCtx, RoutingData, GatherIndx)
from triton_kernels.numerics import InFlexData
from triton_kernels.tensor import convert_layout, wrap_torch_tensor, FP4
from triton_kernels.tensor_details import layout
from triton_kernels.tensor_details.ragged_tensor import make_ragged_tensor_metadata

# gpt-oss-20b MoE, padded dims. K/N here select the 32x256x128 num_warps=2
# shared=25600 specialization -- that is the one that regresses.
E, K, N, TOPK, M = 32, 3072, 6144, 4, 1
dev = "cuda"
torch.manual_seed(0)

# mxfp4 weights + e8m0 block scales, in the layout the kernel expects
vl, vlo = layout.make_default_matmul_mxfp4_w_layout(mx_axis=1)
sl, slo = layout.make_default_matmul_mxfp4_w_scale_layout(mx_axis=1)
w = torch.randint(0, 256, (E, N, K // 2), dtype=torch.uint8, device=dev).mT
s = torch.randint(118, 123, (E, N, K // 32), dtype=torch.uint8, device=dev).mT
w = convert_layout(wrap_torch_tensor(w, dtype=FP4), vl, **vlo)
s = convert_layout(wrap_torch_tensor(s), sl, **slo)
pc = PrecisionConfig(weight_scale=s, flex_ctx=FlexCtx(rhs_data=InFlexData()))

x = torch.randn((M, K), dtype=torch.bfloat16, device=dev)
bias = torch.zeros((E, N), dtype=torch.float32, device=dev)

# routing: the single token goes to TOPK experts
_, idx = torch.topk(torch.randn((M, E), device=dev), TOPK, dim=-1)
idx, src = torch.sort(idx.to(torch.int32).flatten(), stable=True)
hist = torch.histc(idx.float(), bins=E, min=0, max=E - 1).to(torch.int32)
routing = RoutingData(torch.ones((M * TOPK,), dtype=torch.bfloat16, device=dev),
                      hist, E, TOPK, expt_data=make_ragged_tensor_metadata(hist, M * TOPK))
gather = GatherIndx(src.to(torch.int32), torch.argsort(src).to(torch.int32))

run = lambda: matmul_ogs(x, w, bias, routing, gather_indx=gather, precision_config=pc)

for _ in range(20):
    run()
torch.cuda.synchronize()
with profile(activities=[ProfilerActivity.CUDA]) as prof:
    for _ in range(100):
        run()
    torch.cuda.synchronize()

print(f"triton {triton.__version__}  {torch.cuda.get_device_properties(0).gcnArchName}  "
      f"NO_TRACKERS={os.environ.get('NO_TRACKERS','0')}")
for e in prof.key_averages():
    if "matmul_ogs" in e.key:
        us = getattr(e, "self_device_time_total", 0) or getattr(e, "self_cuda_time_total", 0)
        print(f"  {e.key[:52]:<54}{us / max(e.count,1):>8.2f} us x{e.count}")
```

</details>

## Environment

- MI300X, `gfx942:sramecc+:xnack-`
- `vllm/vllm-openai-rocm:v0.28.0`, which ships triton `3.7.1+gitf0b55c07`
- 3.8 arm: triton 3.8 branch @ `669b31acc1dd1b3fd93286afd5db67f65d9f7557`, built from source into that image
- bundled LLVM 22.0.0git

## Additional context

End-to-end on gpt-oss-20b, TP1, concurrency 1, serial runs, 3 warmed passes each
(<=2% within-arm spread):

| triton | trackers | tok/s | TPOT (ms) |
|---|---|---|---|
| 3.6.0 | n/a (flag not in tree) | 121.57 | 8.18 |
| 3.7.1 | ON (upstream default) | 103.95 | 9.58 |
| 3.7.1 | OFF | 125.29 | 7.94 |
| 3.8.0 | OFF (upstream default) | 100.79 | 9.81 |

`TRITON_HIP_USE_ASYNC_COPY=1` does not help this
kernel — ~62% slower in every version/flag combination we measured, and separately it cannot
build the prefill 128-tile on gfx942 (needs 100352 B LDS vs the 65536 B limit).

We have ISA-level and TTGIR-level diffs for both regressions and some idea of what each is
doing; happy to post that detail if useful, but leaving it out here so the reported numbers
stand on their own.
